lib/elixir/lib/macro.ex

```diff
@@ -2475,22 +2477,39 @@ defmodule Macro do
 
     quote do
       unquote(values_ast)
-      {:pipe, unquote(string_asts), Enum.reverse(unquote(values_acc_var))}
+
+      {:pipe,
+       Enum.map(
+         unquote(Macro.escape(asts)),
+         &Macro.to_string_with_colors(&1, unquote(options_var))
+       ), Enum.reverse(unquote(values_acc_var))}
     end
   end
 
   # Any other AST.
-  defp dbg_ast_to_debuggable(ast) do
-    quote do: {:value, unquote(to_string(ast)), unquote(ast)}
+  defp dbg_ast_to_debuggable(ast, options_var) do
+    quote do:
+            {:value,
+             Macro.to_string_with_colors(unquote(Macro.escape(ast)), unquote(options_var)),
+             unquote(ast)}
+  end
+
+  @doc false
+  def to_string_with_colors(ast, options) do
```

Comment:
I think this could be `defp`?

Reply:
Oh, note the goal is to eventually have it a `Macro.to_string/2`. It is just this API is currently soft-deprecated. The goal is to deprecate it at compile-time, and then deprecate it at runtime. That's when we can have `to_string(ast, options \\ [])`. :)

Reply:
> I think this could be defp?

In its current form, this would fail since it is called from quoted code. But my attempt was a bit hacky, I'll probably have to refactor this.

The problem is that [here](https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/macro.ex#L2445-L2446) we have:
- `dbg_ast_to_debuggable` is currently happening at compile time
- it is currently calling `to_string/1` which will need to be changed to `to_string_with_colors`
- `to_string_with_colors` needs to access `options` which is runtime, so I had to move its call to runtime

=> `dbg_ast_to_debuggable` needs at compile time to write the call to `to_string_with_colors` at runtime

Maybe I can manage to move the `to_string_with_colors`  to `__dbg__`, which should contain most of the runtime logic needed.

Reply:
I managed to do it in [54402c6](https://github.com/elixir-lang/elixir/pull/12106/commits/54402c6089811f2c1275dfc5b2b268de13593c74), it seems much better (and now a `defp`).